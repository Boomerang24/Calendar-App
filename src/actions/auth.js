import Swal from "sweetalert2";
import { fetchConToken, fetchSinToken } from "../helpers/fetch"
import { types } from "../types/types";
import { eventLogout } from "./events";


export const startLogin = ( email, password ) => {
    // Al ser tarea async, se utiliza el return con async
    return async( dispatch ) => {

        const resp = await fetchSinToken( 'auth', { email, password }, 'POST' );
        const body = await resp.json();

        if( body.ok ) {
            localStorage.setItem('token', body.token );
            localStorage.setItem('token-init-date', new Date().getTime() ); // Grabar la fecha exacta para saber cuanto tiempo falta para que el token expire

            dispatch( login({
                uid: body.uid,
                name: body.name
            }))
        } else { // body.ok = false
            Swal.fire('Error', body.msg, 'error' );
        }
    }
}

export const startRegister = ( email, password, name ) => {
    return async( dispatch ) => {

        const resp = await fetchSinToken( 'auth/new', { email, password, name }, 'POST' );
        const body = await resp.json();

        if( body.ok ) {
            localStorage.setItem('token', body.token );
            localStorage.setItem('token-init-date', new Date().getTime() ); // Grabar la fecha exacta para saber cuanto tiempo falta para que el token expire

            dispatch( login({
                uid: body.uid,
                name: body.name
            }))
        } else { // body.ok = false
            Swal.fire('Error', body.msg, 'error' );
        }
    }
}

export const startChecking = () => {
    return async( dispatch ) => {

        
        const resp = await fetchConToken( 'auth/renew' );
        const body = await resp.json();

        if( body.ok ) {
            localStorage.setItem('token', body.token );
            localStorage.setItem('token-init-date', new Date().getTime() ); // Grabar la fecha exacta para saber cuanto tiempo falta para que el token expire

            dispatch( login({
                uid: body.uid,
                name: body.name
            }))
        } else {
            dispatch( checkingFinish() );
        }
    }
}

const checkingFinish = () => ({ type: types.authCheckingFinish });

const login = ( user ) => ({
    type: types.authLogin,
    payload: user
});


export const startLogout = () => {
    return ( dispatch )=> {

        localStorage.clear();
        dispatch( eventLogout() );
        dispatch( logout() );
    }
}

const logout = () => ({ type: types.authLogout });
