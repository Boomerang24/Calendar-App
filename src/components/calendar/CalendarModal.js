import React from 'react';

import Modal from 'react-modal';
import { useSelector } from 'react-redux';
import { useDispatch } from 'react-redux';
import { eventClearActive } from '../../actions/events';
import { uiCloseModal } from '../../actions/ui';

import { EventModal } from './EventModal';

const customStyles = {
    content: {
      top: '50%',
      left: '50%',
      right: 'auto',
      bottom: 'auto',
      marginRight: '-50%',
      transform: 'translate(-50%, -50%)',
    },
};

if( process.env.NODE_ENV !== 'test' ){
    Modal.setAppElement('#root'); // setAppElement es el elemento con el que se ligara el modal
}

export const CalendarModal = () => {

    const dispatch = useDispatch();

    const { modalOpen } = useSelector(state => state.ui);

    const closeModal = () => {
        dispatch( uiCloseModal() );
        dispatch( eventClearActive() );
    }

    return (
        <Modal
        className="modal"
        overlayClassName="modal-fondo"
        isOpen={ modalOpen }
        closeTimeoutMS={ 200 }
        onRequestClose={ closeModal }
        style={ customStyles }
        /* Usado para pruebas en modal, por que no tiene espacio para redibujar */
        ariaHideApp={ !process.env.NODE_ENV === 'test' } 
        >
            <EventModal closeModal={ closeModal }/>
            
        </Modal>
    )
}
