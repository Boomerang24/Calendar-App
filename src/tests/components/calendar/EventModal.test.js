import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import moment from 'moment';

import '@testing-library/jest-dom'; // Para autocomplete
import { act } from '@testing-library/react'

import { EventModal } from '../../../components/calendar/EventModal';
import { eventClearActive, eventStartAddNew, eventStartUpdate } from '../../../actions/events';
import { uiCloseModal } from '../../../actions/ui';
import Swal from 'sweetalert2';

jest.mock('../../../actions/events', () => ({
    eventStartUpdate: jest.fn(), // Se simula la funcion
    eventClearActive: jest.fn(),
    eventStartAddNew: jest.fn()
}));

jest.mock('../../../actions/ui', () => ({
    uiCloseModal: jest.fn()
}));

jest.mock('sweetalert2', () => ({
    fire: jest.fn()
}));

const middlewares = [ thunk ];
const mockStore = configureStore( middlewares );

const now = moment().minutes(0).seconds(0).add(1, 'hours');
const endDate = now.clone().add(1, 'hours');

const initState = {
    calendar: {
        events: [],
        activeEvent: {
            title: 'Hola mundo',
            notes: 'Algunas notas',
            start: now.toDate(),
            end: endDate.toDate()
        }
    },
    ui: {
        modalOpen: true
    },
    auth: {
        uid: '1234',
        name: 'Alex'
    }
};

const store = mockStore( initState );
store.dispatch = jest.fn(); // Para simular el dispatch

const closeModal = () => { // Funcion pasada por prompt es simulada
    uiCloseModal()
    eventClearActive()
};

const wrapper = mount(
    <Provider store={ store } >
        <EventModal closeModal={ closeModal }/>
    </Provider>
);

describe('Pruebas en <EventModal />', () => {

    beforeEach(() => {
        jest.clearAllMocks();
    });

    test('debe de mostrarse correctamente', () => {
        
        expect( wrapper ).toMatchSnapshot();
    });
    
    test('debe de llamar la accion de actualizar y cerrar modal', () => {
        
        wrapper.find('form').simulate('submit', {
            preventDefault(){}
        });

        expect( eventStartUpdate ).toHaveBeenCalledWith( initState.calendar.activeEvent );
        expect( uiCloseModal ).toHaveBeenCalled();
    });
    
    test('debe de mostrar error si falta el titulo', () => {
    
        wrapper.find('form').simulate('submit', {
            preventDefault(){}
        });

        expect( wrapper.find('input[name="title"]').hasClass('is-invalid')).toBe( true );

    });

    test('debe de crear un nuevo evento', () => {

        const initState = {
            calendar: {
                events: [],
                activeEvent: null
            },
            ui: {
                modalOpen: true
            },
            auth: {
                uid: '1234',
                name: 'Alex'
            }
        };
        
        const store = mockStore( initState );
        store.dispatch = jest.fn(); // Para simular el dispatch
        
        const wrapper = mount(
            <Provider store={ store } >
                <EventModal closeModal={ closeModal }/>
            </Provider>
        );

        wrapper.find('input[name="title"]').simulate('change', {
            target: {
                name: 'title',
                value: 'Hola pruebas'
            }
        });

        wrapper.find('form').simulate('submit', {
            preventDefault(){}
        });

        expect( eventStartAddNew ).toHaveBeenCalledWith({
            end: expect.anything(),
            start: expect.anything(),
            title: 'Hola pruebas',
            notes: ''
        });

        expect( eventClearActive ).toHaveBeenCalled();
        
    });

    test('debe de validar las fechas', () => {
        
        wrapper.find('input[name="title"]').simulate('change', {
            target: {
                name: 'title',
                value: 'El ultimo test ahora si'
            }
        });

        const hoy = new Date();

        act(() => {
            
            wrapper.find('DateTimePicker').at(1).prop('onChange')(hoy);
        });

        wrapper.find('form').simulate('submit', {
            preventDefault(){}
        });

        expect( Swal.fire ).toHaveBeenCalledWith(
            "Error", "La segunda fecha debe de ser mayor", "error"
        );

    });

});
