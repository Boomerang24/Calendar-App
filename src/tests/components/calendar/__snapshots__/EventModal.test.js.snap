// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <EventModal /> debe de mostrarse correctamente 1`] = `
<div>
  <h1>
     
    Editando evento
     
  </h1>
  <hr />
  <form
    className="container"
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <label>
        Fecha y hora inicio
      </label>
      <div
        className="react-datetime-picker react-datetime-picker--closed react-datetime-picker--enabled form-control"
        onFocus={[Function]}
      >
        <div
          className="react-datetime-picker__wrapper"
        >
          <div
            className="react-datetime-picker__inputGroup"
            onClick={[Function]}
          >
            <input
              max="275760-09-12T19:00"
              min="0001-01-01T00:00"
              name="datetime"
              onChange={[Function]}
              onFocus={[Function]}
              step={60}
              style={
                Object {
                  "position": "absolute",
                  "visibility": "hidden",
                  "zIndex": "-999",
                }
              }
              type="datetime-local"
              value="2022-01-15T01:00"
            />
            <input
              autoComplete="off"
              className="react-datetime-picker__inputGroup__input react-datetime-picker__inputGroup__hour"
              data-input="true"
              inputMode="numeric"
              key="input"
              max={12}
              min={1}
              name="hour12"
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyPress={[Function]}
              onKeyUp={[Function]}
              placeholder="--"
              required={null}
              type="number"
              value="1"
            />
            <span
              className="react-datetime-picker__inputGroup__divider"
            >
              :
            </span>
            Array [
              <span
                className="react-datetime-picker__inputGroup__leadingZero"
                key="leadingZero"
              >
                0
              </span>,
              <input
                autoComplete="off"
                autoFocus={false}
                className="react-datetime-picker__inputGroup__input react-datetime-picker__inputGroup__minute react-datetime-picker__inputGroup__input--hasLeadingZero"
                data-input="true"
                inputMode="numeric"
                key="input"
                max={59}
                min={0}
                name="minute"
                onChange={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyPress={[Function]}
                onKeyUp={[Function]}
                placeholder="--"
                required={null}
                type="number"
                value="0"
              />,
            ]
            <span
              className="react-datetime-picker__inputGroup__divider"
            >
               
            </span>
            <select
              className="react-datetime-picker__inputGroup__input react-datetime-picker__inputGroup__amPm"
              data-input="true"
              data-select="true"
              name="amPm"
              onChange={[Function]}
              required={null}
              value="am"
            >
              <option
                value="am"
              >
                AM
              </option>
              <option
                value="pm"
              >
                PM
              </option>
            </select>
            <span
              className="react-datetime-picker__inputGroup__divider"
            >
               
            </span>
            <input
              autoComplete="off"
              autoFocus={false}
              className="react-datetime-picker__inputGroup__input react-datetime-picker__inputGroup__day"
              data-input="true"
              inputMode="numeric"
              key="input"
              max={31}
              min={1}
              name="day"
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyPress={[Function]}
              onKeyUp={[Function]}
              placeholder="--"
              required={null}
              type="number"
              value="15"
            />
            <span
              className="react-datetime-picker__inputGroup__divider"
            >
              -
            </span>
            Array [
              <span
                className="react-datetime-picker__inputGroup__leadingZero"
                key="leadingZero"
              >
                0
              </span>,
              <input
                autoComplete="off"
                autoFocus={false}
                className="react-datetime-picker__inputGroup__input react-datetime-picker__inputGroup__month react-datetime-picker__inputGroup__input--hasLeadingZero"
                data-input="true"
                inputMode="numeric"
                key="input"
                max={12}
                min={1}
                name="month"
                onChange={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyPress={[Function]}
                onKeyUp={[Function]}
                placeholder="--"
                required={null}
                type="number"
                value="1"
              />,
            ]
            <span
              className="react-datetime-picker__inputGroup__divider"
            >
              -
            </span>
            <input
              autoComplete="off"
              autoFocus={false}
              className="react-datetime-picker__inputGroup__input react-datetime-picker__inputGroup__year"
              data-input="true"
              inputMode="numeric"
              key="input"
              max={275760}
              min={1}
              name="year"
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyPress={[Function]}
              onKeyUp={[Function]}
              placeholder="----"
              required={null}
              step={1}
              type="number"
              value="2022"
            />
          </div>
          <button
            className="react-datetime-picker__clear-button react-datetime-picker__button"
            onClick={[Function]}
            onFocus={[Function]}
            type="button"
          >
            <svg
              className="react-datetime-picker__clear-button__icon react-datetime-picker__button__icon"
              height={19}
              stroke="black"
              strokeWidth={2}
              viewBox="0 0 19 19"
              width={19}
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                x1="4"
                x2="15"
                y1="4"
                y2="15"
              />
              <line
                x1="15"
                x2="4"
                y1="4"
                y2="15"
              />
            </svg>
          </button>
          <button
            className="react-datetime-picker__calendar-button react-datetime-picker__button"
            onClick={[Function]}
            onFocus={[Function]}
            type="button"
          >
            <svg
              className="react-datetime-picker__calendar-button__icon react-datetime-picker__button__icon"
              height={19}
              stroke="black"
              strokeWidth={2}
              viewBox="0 0 19 19"
              width={19}
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                fill="none"
                height="15"
                width="15"
                x="2"
                y="2"
              />
              <line
                x1="6"
                x2="6"
                y1="0"
                y2="4"
              />
              <line
                x1="13"
                x2="13"
                y1="0"
                y2="4"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
    <div>
      <label>
        Fecha y hora fin
      </label>
      <div
        className="react-datetime-picker react-datetime-picker--closed react-datetime-picker--enabled form-control"
        onFocus={[Function]}
      >
        <div
          className="react-datetime-picker__wrapper"
        >
          <div
            className="react-datetime-picker__inputGroup"
            onClick={[Function]}
          >
            <input
              max="275760-09-12T19:00"
              min="2022-01-15T01:00"
              name="datetime"
              onChange={[Function]}
              onFocus={[Function]}
              step={60}
              style={
                Object {
                  "position": "absolute",
                  "visibility": "hidden",
                  "zIndex": "-999",
                }
              }
              type="datetime-local"
              value="2022-01-15T02:00"
            />
            <input
              autoComplete="off"
              className="react-datetime-picker__inputGroup__input react-datetime-picker__inputGroup__hour"
              data-input="true"
              inputMode="numeric"
              key="input"
              max={12}
              min={1}
              name="hour12"
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyPress={[Function]}
              onKeyUp={[Function]}
              placeholder="--"
              required={null}
              type="number"
              value="2"
            />
            <span
              className="react-datetime-picker__inputGroup__divider"
            >
              :
            </span>
            Array [
              <span
                className="react-datetime-picker__inputGroup__leadingZero"
                key="leadingZero"
              >
                0
              </span>,
              <input
                autoComplete="off"
                autoFocus={false}
                className="react-datetime-picker__inputGroup__input react-datetime-picker__inputGroup__minute react-datetime-picker__inputGroup__input--hasLeadingZero"
                data-input="true"
                inputMode="numeric"
                key="input"
                max={59}
                min={0}
                name="minute"
                onChange={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyPress={[Function]}
                onKeyUp={[Function]}
                placeholder="--"
                required={null}
                type="number"
                value="0"
              />,
            ]
            <span
              className="react-datetime-picker__inputGroup__divider"
            >
               
            </span>
            <select
              className="react-datetime-picker__inputGroup__input react-datetime-picker__inputGroup__amPm"
              data-input="true"
              data-select="true"
              name="amPm"
              onChange={[Function]}
              required={null}
              value="am"
            >
              <option
                value="am"
              >
                AM
              </option>
              <option
                value="pm"
              >
                PM
              </option>
            </select>
            <span
              className="react-datetime-picker__inputGroup__divider"
            >
               
            </span>
            <input
              autoComplete="off"
              autoFocus={false}
              className="react-datetime-picker__inputGroup__input react-datetime-picker__inputGroup__day"
              data-input="true"
              inputMode="numeric"
              key="input"
              max={31}
              min={15}
              name="day"
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyPress={[Function]}
              onKeyUp={[Function]}
              placeholder="--"
              required={null}
              type="number"
              value="15"
            />
            <span
              className="react-datetime-picker__inputGroup__divider"
            >
              -
            </span>
            Array [
              <span
                className="react-datetime-picker__inputGroup__leadingZero"
                key="leadingZero"
              >
                0
              </span>,
              <input
                autoComplete="off"
                autoFocus={false}
                className="react-datetime-picker__inputGroup__input react-datetime-picker__inputGroup__month react-datetime-picker__inputGroup__input--hasLeadingZero"
                data-input="true"
                inputMode="numeric"
                key="input"
                max={12}
                min={1}
                name="month"
                onChange={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyPress={[Function]}
                onKeyUp={[Function]}
                placeholder="--"
                required={null}
                type="number"
                value="1"
              />,
            ]
            <span
              className="react-datetime-picker__inputGroup__divider"
            >
              -
            </span>
            <input
              autoComplete="off"
              autoFocus={false}
              className="react-datetime-picker__inputGroup__input react-datetime-picker__inputGroup__year"
              data-input="true"
              inputMode="numeric"
              key="input"
              max={275760}
              min={2022}
              name="year"
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyPress={[Function]}
              onKeyUp={[Function]}
              placeholder="----"
              required={null}
              step={1}
              type="number"
              value="2022"
            />
          </div>
          <button
            className="react-datetime-picker__clear-button react-datetime-picker__button"
            onClick={[Function]}
            onFocus={[Function]}
            type="button"
          >
            <svg
              className="react-datetime-picker__clear-button__icon react-datetime-picker__button__icon"
              height={19}
              stroke="black"
              strokeWidth={2}
              viewBox="0 0 19 19"
              width={19}
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                x1="4"
                x2="15"
                y1="4"
                y2="15"
              />
              <line
                x1="15"
                x2="4"
                y1="4"
                y2="15"
              />
            </svg>
          </button>
          <button
            className="react-datetime-picker__calendar-button react-datetime-picker__button"
            onClick={[Function]}
            onFocus={[Function]}
            type="button"
          >
            <svg
              className="react-datetime-picker__calendar-button__icon react-datetime-picker__button__icon"
              height={19}
              stroke="black"
              strokeWidth={2}
              viewBox="0 0 19 19"
              width={19}
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                fill="none"
                height="15"
                width="15"
                x="2"
                y="2"
              />
              <line
                x1="6"
                x2="6"
                y1="0"
                y2="4"
              />
              <line
                x1="13"
                x2="13"
                y1="0"
                y2="4"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
    <hr />
    <div
      className="form-group"
    >
      <label>
        Titulo y notas
      </label>
      <input
        autoComplete="off"
        className="form-control false "
        name="title"
        onChange={[Function]}
        placeholder="Título del evento"
        type="text"
        value="Hola mundo"
      />
      <small
        className="form-text text-muted"
        id="emailHelp"
      >
        Una descripción corta
      </small>
    </div>
    <div
      className="form-group"
    >
      <textarea
        className="form-control"
        name="notes"
        onChange={[Function]}
        placeholder="Notas"
        rows="5"
        type="text"
        value="Algunas notas"
      />
      <small
        className="form-text text-muted"
        id="emailHelp"
      >
        Información adicional
      </small>
    </div>
    <button
      className="btn btn-outline-primary btn-block"
      type="submit"
    >
      <i
        className="far fa-save"
      />
      <span>
         Guardar
      </span>
    </button>
  </form>
</div>
`;
